File: app/src/main/java/com/song/deviceinfo/utils/LogUtils.java
package com.song.deviceinfo.utils;

import android.os.Handler;
import android.os.HandlerThread;
import android.os.Message;
import android.util.Log;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Created by chensongsong on 2020/6/1.
 */
public class LogUtils {

    private static final int VERBOSE = 1;
    private static final int DEBUG = 2;
    private static final int INFO = 3;
    private static final int WARN = 4;
    private static final int ERROR = 5;
    private static int LEVEL = VERBOSE;
    private static LogUtils.Logger logger = null;
    private static String TAG = Constants.TAG;
    private final static int PRINT_SIZE = 3800;

    public static void init(int level, String tag) {
        LEVEL = level;
        TAG = tag;
    }

    public static void v(String msg) {
        if (LEVEL <= VERBOSE) {
            Log.v(TAG, msg);
            log2sd(TAG, msg);
        }
    }

    public static void d(String msg) {
        if (LEVEL <= DEBUG) {
            Log.d(TAG, msg);
            log2sd(TAG, msg);
        }
    }

    public static void i(String msg) {
        if (LEVEL <= INFO) {
            Log.i(TAG, msg);
            log2sd(TAG, msg);
        }
    }

    public static void w(String msg) {
        if (LEVEL <= WARN) {
            Log.w(TAG, msg);
            log2sd(TAG, msg);
        }
    }

    public static void e(String msg) {
        if (LEVEL <= ERROR) {
            Log.e(TAG, msg);
            log2sd(TAG, msg);
        }
    }

    public static void v(String tag, String msg) {
        if (LEVEL <= VERBOSE) {
            Log.v(tag, msg);
            log2sd(tag, msg);
        }
    }

    public static void d(String tag, String msg) {
        if (LEVEL <= DEBUG) {
            Log.d(tag, msg);
            log2sd(tag, msg);
        }
    }

    public static void i(String tag, String msg) {
        if (LEVEL <= INFO) {
            Log.i(tag, msg);
            log2sd(tag, msg);
        }
    }

    public static void w(String tag, String msg) {
        if (LEVEL <= WARN) {
            Log.w(tag, msg);
            log2sd(tag, msg);
        }
    }

    public static void e(String tag, String msg) {
        if (LEVEL <= ERROR) {
            Log.e(tag, msg);
            log2sd(tag, msg);
        }
    }

    public static void release(String msg) {
        Log.i(TAG, msg);
        log2sd(TAG, msg);
    }

    /**
     * 销毁
     */
    public static void destroy() {
        if (logger != null) {
            logger.destroy();
        }
    }

    private static void log2sd(String tag, String msg) {
        if (logger == null) {
            logger = new LogUtils.Logger();
            logger.init();
            logger.checkLogFile();
        }
        logger.log(tag, msg);
    }

    private static class Logger {
        private HandlerThread thread;
        private Handler handler;
        private SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.CHINA);

        private static final String FILE_NAME = "device_info_log.txt";
        private static final String EXTERNAL_DIR = "song";
        private static final int WHAT_MSG = 0;
        private static final int WHAT_INIT = 1;
        private static final long MAX_FILE_SIZE = 10 * 1024 * 1024; // 10m  10 * 1024 * 1024

        public Logger() {
        }

        public synchronized void init() {
            thread = new HandlerThread("Device Info Thread");
            thread.start();
            handler = new Handler(thread.getLooper()) {

                @Override
                public void handleMessage(Message msg) {
                    super.handleMessage(msg);
                    if (Thread.interrupted()) {
                        return;
                    }
                    if (!SdUtils.isMounted()) {
                        return;
                    }
                    if (msg.what == WHAT_MSG) {
                        LogUtils.Logger.Item item = (LogUtils.Logger.Item) msg.obj;
                        write(build(sdf, item.millis, item.tag, item.message));
                    } else if (msg.what == WHAT_INIT) {
                        deleteCauseExceedMaxSize();
                    }
                }
            };
        }

        private static class Item {
            public long millis;
            public String tag;
            public String message;
        }

        public synchronized void log(String tag, String msg) {

            Message message = handler.obtainMessage();
            message.what = WHAT_MSG;

            LogUtils.Logger.Item item = new LogUtils.Logger.Item();
            item.millis = System.currentTimeMillis();
            item.tag = tag;
            item.message = msg;
            message.obj = item;

            handler.sendMessage(message);
        }

        public synchronized void checkLogFile() {
            Message message = handler.obtainMessage();
            message.what = WHAT_INIT;
            handler.sendMessage(message);
        }

        public synchronized void destroy() {
            if (thread != null && handler != null) {
                handler.removeMessages(WHAT_MSG);
                handler.removeMessages(WHAT_INIT);

                thread.quit();
                thread = null;
            }
        }

        private static String externalDirPath() {
            return SdUtils.getDirPath() + File.separator + EXTERNAL_DIR;
        }

        private static boolean deleteCauseExceedMaxSize() {
            File dir = new File(externalDirPath());
            if (!dir.exists()) {
                return false;
            }
            File f = new File(dir, FILE_NAME);
            if (!f.exists()) {
                return false;
            }
            if (f.length() < MAX_FILE_SIZE) {
                return false;
            }
            return f.delete();
        }

        private void write(String content) {

            File dir = new File(externalDirPath());
            if (!dir.exists()) {
                dir.mkdirs();
            }

            BufferedOutputStream out = null;
            try {
                out = new BufferedOutputStream(new FileOutputStream(new File(dir, FILE_NAME), true));
                out.write(content.getBytes("utf-8"));
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                if (out != null) {
                    try {
                        out.close();
                    } catch (IOException e) {
                    }
                }
            }
        }

        private static String build(SimpleDateFormat sdf, long millis, String tag, String msg) {
            StringBuilder sb = new StringBuilder();
            sb.append(sdf.format(new Date(millis)));
            sb.append('\t');
            sb.append(tag);
            sb.append('\n');
            sb.append(msg);
            sb.append('\n');
            return sb.toString();
        }
    }

    /**
     * 打印超长字符串
     *
     * @param data
     */
    public static void printLongString(String tag, String data) {

        int len = data.length();
        if (len > PRINT_SIZE) {
            int n = 0;
            while ((len - n) > PRINT_SIZE) {
                String s = data.substring(n, PRINT_SIZE);
                Log.e(TAG, tag + s);
                n += PRINT_SIZE;
            }
            Log.e(TAG, tag + data.substring(n));
        } else {
            Log.e(TAG, tag + data);
        }
    }
}

File: app/src/main/java/com/song/deviceinfo/utils/LogUtils.kt
package com.song.deviceinfo.utils

import android.content.Context
import android.os.Environment
import android.os.Handler
import android.os.HandlerThread
import android.os.Message
import android.util.Log
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Created by chensongsong on 2020/6/1.
 */
object LogUtils {
    private const val VERBOSE = 1
    private const val DEBUG = 2
    private const val INFO = 3
    private const val WARN = 4
    private const val ERROR = 5
    private var LEVEL = VERBOSE
    private var logger: Logger? = null
    private var TAG = Constants.TAG
    private const val PRINT_SIZE = 3800
    
    fun init(level: Int, tag: String) {
        LEVEL = level
        TAG = tag
    }
    
    fun v(msg: String) {
        if (LEVEL <= VERBOSE) {
            Log.v(TAG, msg)
            log2sd(TAG, msg)
        }
    }
    
    @JvmStatic
    fun d(msg: String) {
        if (LEVEL <= DEBUG) {
            Log.d(TAG, msg)
            log2sd(TAG, msg)
        }
    }
    
    fun i(msg: String) {
        if (LEVEL <= INFO) {
            Log.i(TAG, msg)
            log2sd(TAG, msg)
        }
    }
    
    fun w(msg: String) {
        if (LEVEL <= WARN) {
            Log.w(TAG, msg)
            log2sd(TAG, msg)
        }
    }
    
    @JvmStatic
    fun e(msg: String) {
        if (LEVEL <= ERROR) {
            Log.e(TAG, msg)
            log2sd(TAG, msg)
        }
    }
    
    fun v(tag: String, msg: String) {
        if (LEVEL <= VERBOSE) {
            Log.v(tag, msg)
            log2sd(tag, msg)
        }
    }
    
    fun d(tag: String, msg: String) {
        if (LEVEL <= DEBUG) {
            Log.d(tag, msg)
            log2sd(tag, msg)
        }
    }
    
    fun i(tag: String, msg: String) {
        if (LEVEL <= INFO) {
            Log.i(tag, msg)
            log2sd(tag, msg)
        }
    }
    
    fun w(tag: String, msg: String) {
        if (LEVEL <= WARN) {
            Log.w(tag, msg)
            log2sd(tag, msg)
        }
    }
    
    fun e(tag: String, msg: String) {
        if (LEVEL <= ERROR) {
            Log.e(tag, msg)
            log2sd(tag, msg)
        }
    }
    
    @JvmStatic
    fun release(msg: String) {
        Log.i(TAG, msg)
        log2sd(TAG, msg)
    }
    
    /**
     * 销毁
     */
    fun destroy() {
        logger?.destroy()
    }
    
    private fun log2sd(tag: String, msg: String) {
        logger?.log(tag, msg)
    }
    
    fun initLogger(context: Context) {
        if (logger == null) {
            logger = Logger()
            logger?.init(context)
            logger?.checkLogFile()
        }
    }
    
    /**
     * 打印超长字符串
     *
     * @param data
     */
    @JvmStatic
    fun printLongString(data: String) {
        val len = data.length
        if (len > PRINT_SIZE) {
            var n = 0
            while ((len - n) > PRINT_SIZE) {
                val s = data.substring(n, n + PRINT_SIZE)
                Log.i(TAG, s)
                n += PRINT_SIZE
            }
            Log.i(TAG, data.substring(n))
        } else {
            Log.i(TAG, data)
        }
    }
    
    private class Logger {
        private var thread: HandlerThread? = null
        private var handler: Handler? = null
        private val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.CHINA)
        private var context: Context? = null
        
        @Synchronized
        fun init(context: Context) {
            this.context = context.applicationContext
            thread = HandlerThread("Device Info Thread")
            thread?.start()
            handler = object : Handler(thread?.looper!!) {
                override fun handleMessage(msg: Message) {
                    super.handleMessage(msg)
                    if (Thread.interrupted()) {
                        return
                    }
                    if (msg.what == WHAT_MSG) {
                        val item = msg.obj as Item
                        write(build(sdf, item.millis, item.tag, item.message))
                    } else if (msg.what == WHAT_INIT) {
                        deleteCauseExceedMaxSize(context)
                    }
                }
            }
        }
        
        private class Item {
            var millis: Long = 0
            var tag: String? = null
            var message: String? = null
        }
        
        @Synchronized
        fun log(tag: String?, msg: String?) {
            val message = handler!!.obtainMessage()
            message.what = WHAT_MSG
            
            val item = Item()
            item.millis = System.currentTimeMillis()
            item.tag = tag
            item.message = msg
            message.obj = item
            
            handler!!.sendMessage(message)
        }
        
        @Synchronized
        fun checkLogFile() {
            val message = handler!!.obtainMessage()
            message.what = WHAT_INIT
            handler!!.sendMessage(message)
        }
        
        @Synchronized
        fun destroy() {
            if (thread != null && handler != null) {
                handler!!.removeMessages(WHAT_MSG)
                handler!!.removeMessages(WHAT_INIT)
                
                thread!!.quit()
                thread = null
            }
            context = null
        }
        
        fun write(content: String) {
            context?.let { ctx ->
                val dir = File(ctx.getExternalFilesDir(null), "logs")
                if (!dir.exists()) {
                    dir.mkdirs()
                }
                
                var out: BufferedOutputStream? = null
                try {
                    out = BufferedOutputStream(FileOutputStream(File(dir, FILE_NAME), true))
                    out.write(content.toByteArray(charset("utf-8")))
                } catch (e: Exception) {
                    e.printStackTrace()
                } finally {
                    out?.close()
                }
            }
        }
        
        companion object {
            private const val FILE_NAME = "device_info_log.txt"
            private const val WHAT_MSG = 0
            private const val WHAT_INIT = 1
            private const val MAX_FILE_SIZE = 10L * 1024 * 1024 // 10MB
            
            private fun build(sdf: SimpleDateFormat, millis: Long, tag: String?, message: String?): String {
                return "${sdf.format(Date(millis))} $tag: $message\n"
            }
            
            private fun deleteCauseExceedMaxSize(context: Context) {
                context.let { ctx ->
                    val dir = File(ctx.getExternalFilesDir(null), "logs")
                    if (!dir.exists()) {
                        return
                    }
                    val file = File(dir, FILE_NAME)
                    if (file.exists() && file.length() > MAX_FILE_SIZE) {
                        file.delete()
                    }
                }
            }
        }
    }
}
